src/main/java/sparksoniq/spark/DataFrameUtils.java

```diff
@@ -225,27 +225,26 @@ public static String getGroupbyProjectSQL(
             }
             if (applyCount) {
                 queryColumnString.append("count(`");
-            } else if (applyDistinct) {
-                queryColumnString.append(serializerUdfName);
-                queryColumnString.append("(");
-                queryColumnString.append("array(");
-                queryColumnString.append("first(`");
             } else {
                 queryColumnString.append(serializerUdfName);
                 queryColumnString.append("(");
-                queryColumnString.append("collect_list(`");
+                if (applyDistinct) {
+                    queryColumnString.append("array(");
+                    queryColumnString.append("first(`");
+                } else {
+                    queryColumnString.append("collect_list(`");
```

Comment:
I think this should not be in an else statement

Reply:
Actually this is intentional -- first and collect_list are both aggregation operators. This is also why I think performance may improve significantly, because there is no need to collect anything any more.

We only take the first value and put it into an array.

Reply:
Thanks for the explanation, sounds great !